src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/FileTypeNameAnalyzer.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Immutable;
+using System.IO;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Azure.ClientSdk.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class FileTypeNameAnalyzer : DiagnosticAnalyzer
+    {
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+
+            context.RegisterCompilationStartAction(
+                analysisContext => {
+                    analysisContext.RegisterSymbolAction(
+                        symbolAnalysisContext => {
+                            var symbol = symbolAnalysisContext.Symbol;
+
+                            if (symbol.DeclaredAccessibility != Accessibility.Public)
```

Comment:
Why not verify internal types too?  I've always had a hard time navigating around pipeline internals because the files are often named after the patterns from three iterations ago.

Reply:
Mostly because of the side effect that you've mentioned below but maybe it's fine and we should do single type per file even for internals.